sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -126,14 +129,30 @@ class CostBasedOptimizer(conf: RapidsConf) extends Logging {
       totalGpuCost = totalCpuCost
     }
 
-    if (!plan.canThisBeReplaced) {
+    if (!plan.canThisBeReplaced || mustThisBeReplaced(plan)) {
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
     (totalCpuCost, totalGpuCost)
   }
 
+  private def mustThisBeReplaced(plan: SparkPlanMeta[_]): Boolean = {
```

Comment:
Does this really mean something more like, "planning for this node cannot be changed"?  Just because we find a shuffle or other node involving data transfer in the original plan, that doesn't mean we always are replacing those nodes with GPU versions, correct?

Reply:
Good point. This method is really just determining whether the operator is consuming output from a query stage, so I have renamed it `consumesQueryStage` now.